src/github.com/getlantern/peerscanner/main.go

```diff
@@ -0,0 +1,207 @@
+
+// loadHosts loads the initial list of hosts based on what's in CloudFlare's
+// DNS at startup.
+func loadHosts() (map[string]*host, error) {
+	log.Debug("Loading existing hosts from CloudFlare ...")
+
+	recs, err := cfutil.GetAllRecords()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to load hosts: %v", err)
+	}
+
+	// Keep track of different groups of hosts
+	groups := map[string]map[string]*cloudflare.Record{
+		RoundRobin: make(map[string]*cloudflare.Record),
+		Fallbacks:  make(map[string]*cloudflare.Record),
+		Peers:      make(map[string]*cloudflare.Record),
+	}
+
+	addToGroup := func(name string, r cloudflare.Record) {
+		log.Tracef("Adding to %v: %v", name, r.Value)
+		groups[name][r.Value] = &r
+	}
+
+	// Build map of existing hosts
+	hosts := make(map[string]*host)
+
+	addHost := func(r cloudflare.Record) {
+		h := newHost(r.Name, r.Value, &r)
+		hosts[h.ip] = h
+	}
+
+	// Look through all records to find peers, fallbacks and groups
+	for _, r := range recs {
+		if isFallback(r.Name) {
+			log.Tracef("Adding fallback: %v", r.Name)
+			addHost(r)
+		} else if isPeer(r.Name) {
+			log.Tracef("Adding peer: %v", r.Name)
+			addHost(r)
+		} else if r.Name == RoundRobin {
+			addToGroup(RoundRobin, r)
+		} else if r.Name == Fallbacks {
+			addToGroup(Fallbacks, r)
+		} else if r.Name == Peers {
+			addToGroup(Peers, r)
+		} else {
+			log.Tracef("Unrecognized record: %v", r.FullName)
+		}
+	}
+
+	// Update hosts with group info
+	for _, h := range hosts {
+		for _, hg := range h.groups {
+			g, found := groups[hg.subdomain]
+			if found {
+				hg.existing = g[h.ip]
+				delete(g, h.ip)
+			}
+		}
+	}
+
+	// Remove items from rotation that don't have a corresponding host
+	var wg sync.WaitGroup
+	for k, g := range groups {
+		for _, r := range g {
+			wg.Add(1)
+			go removeFromRotation(&wg, k, r)
+		}
+	}
+	wg.Wait()
+
+	// Start hosts
+	for _, h := range hosts {
+		h.run()
+	}
+
+	return hosts, nil
+}
+
+func removeFromRotation(wg *sync.WaitGroup, k string, r *cloudflare.Record) {
+	log.Debugf("%v in %v is missing host, removing from rotation", r.Value, k)
+	err := cfutil.DestroyRecord(r)
+	if err != nil {
+		log.Debugf("Unable to remove %v from %v: %v", r.Value, k, err)
+	}
+	wg.Done()
+}
+
+func getOrCreateHost(name string, ip string) *host {
+	hostsMutex.Lock()
+	defer hostsMutex.Unlock()
+
+	h := hostsByIp[ip]
+	if h == nil {
+		h := newHost(name, ip, nil)
+		hostsByIp[ip] = h
+		go h.run()
+		return h
+	}
+	h.reset(name)
+	return h
+}
+
+func getHostByIp(ip string) *host {
+	hostsMutex.Lock()
+	defer hostsMutex.Unlock()
+	return hostsByIp[ip]
+}
+
+func isPeer(name string) bool {
+	// We just check the length of the subdomain here, which is the unique
+	// peer GUID. While it's possible something else could have a subdomain
+	// this long, it's unlikely.
+	// We also accept anything with a name beginning with peer- as a peer
+	return len(name) == 32 || strings.Index(name, "peer-") == 0
```

Comment:
`strings.Index(name, "peer-") == 0` works but we could also use `strings.HasPrefix("name", "peer-")`.